org/mozilla/jss/PK11Finder.c

```diff
@@ -1587,8 +1590,24 @@ SECStatus verifyCertificateNow(JNIEnv *env, jobject self, jstring nickString,
     /* 0 for certificateUsage in call to CERT_VerifyCertificateNow will
      * retrieve the current valid usage into currUsage
      */
-        rv = CERT_VerifyCertificateNow(CERT_GetDefaultCertDB(), cert,
-            checkSig, certificateUsage, NULL, currUsage );
+        if( ocspPolicy == OCSP_LEAF_AND_CHAIN_POLICY) {
+            rv = JSSL_verifyCertPKIX( cert, certificateUsage,
+                     NULL /* pin arg */, ocspPolicy, NULL, currUsage);
+
+            /* we need to do this just to get the cert usages, the pkix version
+               doesn't seem to honor the method to get the usages as of yet.
+               Let the PKIX call only determine the final fate.
+            */
+            if(rv == SECSuccess) {
+                CERT_VerifyCertificateNow(CERT_GetDefaultCertDB(), cert,
```

Comment:
I wonder if there's a lighter weight function to call to get certificate usages if we don't care about the rest of the work `CERT_VerifyCertificateNow` does.

Reply:
Yes, in this crazy bit of code, we don't care what CERT_VerifyCertNow says , we just need it to give us the correct currUsage. We pass it into the new PKIX call because it allegedly should be providing us with the info, but appears not to work. I just left the call in there so we can switch to it if and when it ever actually works. I poked around to see if there was a lighter weight way to get that info, but ran out of time. In the normal case there will be caching going on so the extra call won't be that bad.  The bit of simplification mentioned sounds reasonable, but I was running out of time. I wonder if I can still do it... Let me take a look and think.

Reply:
Ah, the part about simplifying the above if branch where we are forced to call CERT_VerifyCertNow just to get currUsage. We could put this inside the PKIX method, but as you see the CERT style function takes some args that make no sense for the PKIX method, such as the cert db and a checkSig boolean. I'm open to opinions though.

Reply:
Ah I see that now -- yes, `CERT_GetDefaultCertDB()` and `checkSig` aren't passed to `JSSL_verifyCertPKIX`. Fair point here.

I'm still not sure why we don't care about the retval on this though. In the event we want to validate OCSP responses (which IMO, is a "global" configuration--you enable it on CryptoManager and set it instance wide--vs say, a per-call configuration (as an additional argument, what type of OCSP response checking you want to do))... ...don't we still want to verify the certificates' signature/expiration/etc.? Perhaps `PKIX_verifyCert` does that already though....

Correct my understanding of NSS if it is wrong, but doesn't `CERT_VerifyCertificateNow` check the signature conditionally (if `checkSig` is true) and `JSSL_verifyCertPKIX` always check signature (or rather, `CERT_PKIXVerifyCert` as that appears to do most of the heavy lifting)? I guess the issue isn't `JSSL_*` returns True and `CERT_VerifyCertificateNow` doesn't, but `checkSig == false` and `JSSL_*` returns `false` but `CERT_VerifyCertificateNow` would've returned true.

Reply:
As for the first point,in the one scenario supported, we rely on the answer that the PKIX call gives us for the actual result yes or no. So I figured to just go with that and use the CERT call to get the stubborn usages value out. Supposedly the PKIX library is supposed to give more accurate answers due to things like complicated trust trees and such.

As for the point about check sig, I'm not sure what lib PKIX is doing. Maybe it does that for you already. Anway, if it makes any sense, I wanted to use PKIX only for the one specific scenario when we want AIA ocsp checking, and leaving all other scenarios the same. Of course later we can move completely to PKIX at some point for everything and simplify the logic.

Reply:
Ah I understand now and am fine with this. The new call is a replacement for the old one which performs the same tasks.